plugins/pulp_rpm/plugins/importers/yum/purge.py

```diff
@@ -286,3 +289,209 @@ def remove_unit_duplicate_nevra(unit, repo):
+        _logger.info('Purging duplicate NEVRA can take significantly longer in versions of '
+                     'mongodb lower than 2.6. Consider upgrading mongodb if cleaning duplicate '
+                     'packages takes an unreasonably long time.')
+        return _duplicate_key_id_generator_mapreduce(unit, fields)
+
+
+def _duplicate_key_id_generator_aggregation(unit, fields):
+    """Superior Aggregation version of duplicate nevra unit ID generator for mongo 2.6+
+
+    In order to find potential duplicate nevras, a mongo aggregation pipeline is employed
+    to sort a unit's entire collection based on non-checksum unit key fields. Doing the sorting
+    purely as a normal QuerySet overflow's mongo's sort stage bugger, and doing it purely with
+    python uses too much memory. Running things through the aggregation pipelines results in
+    an iterable that makes it relatively simple to spit out potential duplicate unit IDs without
+    breaking either mongo or python. Because the resulting dataset can be rather large, it is
+    important to be able to use a mongo db cursor to iterate over the results, which is sadly
+    unsupported in 2.4. In that case, the mapreduce method below must be used.
+
+    :param unit: The unit whose NEVRA should be removed
+    :type unit: ContentUnit
+    :param fields: list of fields that define a given unit type's NEVRA fields
+    :type fields: list
+    """
+
+    # create the aggregation params by zipping the fields with 1
+    # It is a happy coincidence that the two dicts are the same values
+    pipeline_opts = dict(zip(fields, repeat(1)))
+
+    # for $sort, this indicates ascending sort for nevra fields
+    sort = {'$sort': pipeline_opts}
+
+    # for $project, this indicates what fields to include in the result
+    project = {'$project': pipeline_opts}
+
+    # When aggregating over hundreds of thousands of packages, mongo can overflow
+    # To prevent this, mongo needs to be allowed to temporarily use the disk for this transaction
+    aggregation = unit.objects.aggregate(sort, project, allowDiskUse=True)
+
+    # loop state tracking vars
+    previous_nevra = None
+    previous_pkg_id = None
+    yielding_ids = None
+
+    for pkg in aggregation:
+        # strip the checksums and mongo metadata so they don't get used in equality checks
+        current_nevra = tuple(pkg[field] for field in fields)
+        # current nevra matches previous: this is a duplicate nevra
+        if current_nevra == previous_nevra:
+            if yielding_ids is None:
+                # The current nevra is a duplicate but yielding_ids is None, which means
+                # this iteration is the first to detect a duplicate for this nevra.
+                # Initialize the list of ids to yield with the *previous* id,
+                # since it was the first unit seen with the current nevra
+                yielding_ids = [previous_pkg_id]
```

Comment:
What if you intialized yielding_ids to [] instead of None so you can always append below? Of course, you'd also need to reset it to [] when moving on to the next nevra, but it might make the code slightly smaller and let you have less 'splainin' to do.